File: PHP/238. Product of Array Except Self.php
<?php

/**
 *  @noinspection PhpIllegalPsrClassPathInspection, PhpMultipleClassDeclarationsInspection
 */
class Solution
{
    /**
     * @param  int[]  $nums
     * @return int[]
     *
     * @noinspection PhpUnused
     */
    public function productExceptSelf(array $nums): array
    {
        $length = count($nums);
        $result = array_fill(0, $length, 1);

        $num = 1;
        for ($i = $length - 1; $i >= 0; $i--) {
            $result[$i] = $num;
            $num *= $nums[$i];
        }

        $num = 1;
        for ($i = 0; $i < $length; $i++) {
            $result[$i] *= $num;
            $num *= $nums[$i];
        }

        return $result;
    }
}

File: Python/238. Product of Array Except Self.py
from typing import List


# noinspection PyMethodMayBeStatic
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        count = len(nums)
        result = [1] * count

        num = 1
        for i in range(count - 1, -1, -1):
            result[i] = num
            num *= nums[i]

        num = 1
        for i in range(0, count):
            result[i] *= num
            num *= nums[i]

        return result
